fix: use temporary dir for package manager files

synackd opened the pull request:
Instead of relying on a fixed directory in `/tmp` that may or may not exist to write package manager logs/cache to, use the tempfile library to create a temporary directory within the container for these at runtime.

Also, move the functions in installer.py to a new Installer class in order to store and reuse the buildah container and mount names, the package manager name, as well as the temporary directory which is created on init.

This is to avoid errors from DNF complaining that `/tmp/dnf_test` doesn't exist or that it failed to open `hawkey.log` because the log directory doesn't exist.

synackd commented:
I've tested this with two DNF image configs: one from scratch and one using an existing parent. More testing would be helpful, though.

synackd commented on src/installer.py:
```diff
@@ -1,194 +1,210 @@
+
+        # Create temporary directory for logs, cache, etc. for package manager
+        os.makedirs(os.path.join(mname, "tmp"))
+        self.tdir = tempfile.mkdtemp(dir=os.path.join(self.mname, "tmp"), prefix="image-build-")
```
Right now the tempdir is being created within the container being created, which gets umounted/deleted if errors occur. It just occurred to me that it might be useful to examine package manager logs should something go wrong and so it might make more sense to create the tempdir outside of the container on the system (whether host system or container) where `image-build` is being run. That way, the logs could be examined even if errors occur and the container that `image-build` creates gets deleted.

synackd commented:
`image-build` should now print the path to the temporary directory it creates:
```
$ podman run -it --rm --device /dev/fuse -v ./config.yaml:/home/builder/config.yaml:Z localhost/image-build:test bash
[root@d5657938e178 ~]# image-build --config config.yaml --log-level DEBUG
...
INFO - Installer: Temporary directory for dnf created at /var/tmp/image-build-0xg6_747
...
```

These files can be inspected afterward:

```
[root@d5657938e178 ~]# ls /var/tmp/image-build-0xg6_747
dnf
[root@d5657938e178 ~]# ls /var/tmp/image-build-0xg6_747/dnf/log
dnf.librepo.log  dnf.log  dnf.rpm.log  hawkey.log
[root@d5657938e178 ~]# tail /var/tmp/image-build-0xg6_747/dnf/log/dnf.log
2025-02-12T16:50:48+0000 DEBUG DNF version: 4.7.0
2025-02-12T16:50:48+0000 DDEBUG Command: dnf --setopt=reposdir=/home/builder/.local/share/containers/storage/overlay/ff8bbb2f5a4ba649b19f731c1ccd55dd41e9810a1c881e9b9a32eeace439d99f/merged/etc/yum.repos.d --setopt=logdir=/var/tmp/image-build-0xg6_747/dnf/log --setopt=cachedir=/var/tmp/image-build-0xg6_747/dnf/cache config-manager --save --add-repo http://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/ 
2025-02-12T16:50:48+0000 DDEBUG Installroot: /
2025-02-12T16:50:48+0000 DDEBUG Releasever: 8
2025-02-12T16:50:48+0000 DEBUG cachedir: /var/tmp/image-build-0xg6_747/dnf/cache
2025-02-12T16:50:48+0000 DDEBUG Base command: config-manager
2025-02-12T16:50:48+0000 DDEBUG Extra commands: ['--setopt=reposdir=/home/builder/.local/share/containers/storage/overlay/ff8bbb2f5a4ba649b19f731c1ccd55dd41e9810a1c881e9b9a32eeace439d99f/merged/etc/yum.repos.d', '--setopt=logdir=/var/tmp/image-build-0xg6_747/dnf/log', '--setopt=cachedir=/var/tmp/image-build-0xg6_747/dnf/cache', 'config-manager', '--save', '--add-repo', 'http://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/']
2025-02-12T16:50:48+0000 INFO Adding repo from: http://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/
2025-02-12T16:50:48+0000 DDEBUG Cleaning up.
2025-02-12T16:50:48+0000 DDEBUG Plugins were unloaded.
```

Things get a little funky with permissions when trying to mount a directory to `/var/tmp` within the container since the container runs with `buildah unshare`.

```
$ id -u
5796
$ podman run -it --rm --device /dev/fuse -v ~/image-tmp:/var/tmp/:rw,Z,U -v ./base-8.10.yaml:/home/builder/config.yaml:Z localhost/image-build:test image-build --config config.yaml --log-level DEBUG
$ ls -ld ~/image-tmp/ | cut -f3 -d' '
297609
```

Since there are other ways to view the files if one is not root to change the permissions. I won't consider this in the scope of this PR.

davidallendj reviewed (approved):
I tested this on our Redondo cluster and was able to build a rocky 9 image.

alexlovelltroy commented:
Can we double check that the documentation is still correct at https://github.com/OpenCHAMI/image-builder ?  If not, let's update it and merge.  It seems like the errors I was seeing are likely user error.

synackd commented:
I updated the README to add comments to and update the example config for the base layer. I also updated the publish section and updated the run command for the base layer to use podman instead of running bare-metal. I don't have an example of an Ansible-type layer, but can add that too if someone has one (@travisbcotton ?).

synackd commented:
Added example config for Ansible-type layer. This should be ready for review again.

alexlovelltroy reviewed (approved):
LGTM

tested locally
